pubsub/pubsub_test.go

```diff
@@ -0,0 +1,397 @@
+	return []byte(val)
+}
+
+func (t *testResponseMarshaller) Unmarshal(val []byte) (string, error) {
+	return string(val), nil
+}
+
+func createGroup(ctx context.Context, t *testing.T, streamName, groupName string, client redis.UniversalClient) {
+	t.Helper()
+	_, err := client.XGroupCreateMkStream(ctx, streamName, groupName, "$").Result()
+	if err != nil {
+		t.Fatalf("Error creating stream group: %v", err)
+	}
+}
+
+func destroyGroup(ctx context.Context, t *testing.T, streamName, groupName string, client redis.UniversalClient) {
+	t.Helper()
+	_, err := client.XGroupDestroy(ctx, streamName, groupName).Result()
+	if err != nil {
+		t.Fatalf("Error creating stream group: %v", err)
+	}
+}
+
+type configOpt interface {
+	apply(consCfg *ConsumerConfig, prodCfg *ProducerConfig)
+}
+
+type disableReproduce struct{}
+
+func (e *disableReproduce) apply(_ *ConsumerConfig, prodCfg *ProducerConfig) {
+	prodCfg.EnableReproduce = false
+}
+
+func producerCfg() *ProducerConfig {
+	return &ProducerConfig{
+		EnableReproduce:      DefaultTestProducerConfig.EnableReproduce,
+		CheckPendingInterval: DefaultTestProducerConfig.CheckPendingInterval,
+		KeepAliveTimeout:     DefaultTestProducerConfig.KeepAliveTimeout,
+		CheckResultInterval:  DefaultTestProducerConfig.CheckResultInterval,
+	}
+}
+
+func consumerCfg() *ConsumerConfig {
+	return &ConsumerConfig{
+		ResponseEntryTimeout: DefaultTestConsumerConfig.ResponseEntryTimeout,
+		KeepAliveTimeout:     DefaultTestConsumerConfig.KeepAliveTimeout,
+	}
+}
+
+func newProducerConsumers(ctx context.Context, t *testing.T, opts ...configOpt) (*Producer[string, string], []*Consumer[string, string]) {
+	t.Helper()
+	redisURL := redisutil.CreateTestRedis(ctx, t)
+	prodCfg, consCfg := producerCfg(), consumerCfg()
+	prodCfg.RedisURL, consCfg.RedisURL = redisURL, redisURL
+	streamName := uuid.NewString()
+	groupName := fmt.Sprintf("group_%s", streamName)
+	prodCfg.RedisGroup, consCfg.RedisGroup = groupName, groupName
+	prodCfg.RedisStream, consCfg.RedisStream = streamName, streamName
+	for _, o := range opts {
+		o.apply(consCfg, prodCfg)
+	}
+	producer, err := NewProducer[string, string](prodCfg, &testRequestMarshaller{}, &testResponseMarshaller{})
+	if err != nil {
+		t.Fatalf("Error creating new producer: %v", err)
+	}
+
+	var consumers []*Consumer[string, string]
+	for i := 0; i < consumersCount; i++ {
+		c, err := NewConsumer[string, string](ctx, consCfg, &testRequestMarshaller{}, &testResponseMarshaller{})
+		if err != nil {
+			t.Fatalf("Error creating new consumer: %v", err)
+		}
+		consumers = append(consumers, c)
+	}
+	createGroup(ctx, t, streamName, groupName, producer.client)
+	t.Cleanup(func() {
+		destroyGroup(ctx, t, streamName, groupName, producer.client)
+		var keys []string
+		for _, c := range consumers {
+			keys = append(keys, c.heartBeatKey())
+		}
+		if _, err := producer.client.Del(ctx, keys...).Result(); err != nil {
+			t.Fatalf("Error deleting heartbeat keys: %v\n", err)
```

Comment:
Possibly log, but I don't think this should cause the test to fail.. I don't know that even calling it here can make the test fail.
Specifically in ClimingOwnership - some of the heartbeat values should not exist (will that result in an error?)